MISampleApp/VaultProbe/VaultProbe.cs

```diff
@@ -0,0 +1,221 @@
+                {
+                    response = String.Format($"Non-critical error encountered retrieving secret '{secret}' in vault '{vault}': {secretResponse.Response.ReasonPhrase} ({secretResponse.Response.StatusCode})");
+                }
+            }
+            catch (Microsoft.Rest.ValidationException ve)
+            {
+                response = String.Format($"encountered REST validation exception 0x{ve.HResult.ToString("X")} trying to access '{secret}' in vault '{vault}' from {ve.Source}: {ve.Message}");
+            }
+            catch (KeyVaultErrorException kvee)
+            {
+                response = String.Format($"encountered KeyVault exception 0x{kvee.HResult.ToString("X")} trying to access '{secret}' in vault '{vault}': {kvee.Response.ReasonPhrase} ({kvee.Response.StatusCode})");
+            }
+            catch (Exception ex)
+            {
+                // handle generic errors here
+                response = String.Format($"encountered exception 0x{ex.HResult.ToString("X")} trying to access '{secret}' in vault '{vault}': {ex.Message}");
+            }
+
+            Log(LogLevel.Info, response);
+
+            return response;
+        }
+
+        /// <summary>
+        /// KV authentication callback, using the application's managed identity.
+        /// </summary>
+        /// <param name="authority"></param>
+        /// <param name="resource"></param>
+        /// <param name="scope"></param>
+        /// <returns>Access token</returns>
+        public async Task<string> AuthenticationCallbackAsync(string authority, string resource, string scope)
+        {
+            Log(LogLevel.Verbose, $"authentication callback invoked with: auth: '{authority}', resource: '{resource}', scope: '{scope}'");
+
+            var encodedResource = HttpUtility.UrlEncode(resource);
+
+            // first check the cache; use the resource as the caching key
+            ManagedIdentityTokenResponse tokenResponse;
+            if (responseCache.TryGetCachedItem(encodedResource, out tokenResponse))
+            {
+                Log(LogLevel.Verbose, $"cache hit for key '{encodedResource}'");
+
+                return tokenResponse.AccessToken;
+            }
+
+            Log(LogLevel.Verbose, $"cache miss for key '{encodedResource}'");
+
+            var requestUri = $"{config.ManagedIdentityEndpoint}?api-version={config.ManagedIdentityApiVersion}&resource={encodedResource}";
+            Log(LogLevel.Verbose, $"request uri: {requestUri}");
+
+            var requestMessage = new HttpRequestMessage(HttpMethod.Get, requestUri);
+            requestMessage.Headers.Add(config.ManagedIdentityAuthenticationHeader, config.ManagedIdentityAuthenticationCode);
+            Log(LogLevel.Verbose, $"added header '{config.ManagedIdentityAuthenticationHeader}': '{config.ManagedIdentityAuthenticationCode}'");
+
+            var response = await httpClient.SendAsync(requestMessage)
+                .ConfigureAwait(false);
```

Comment:
We should add here support for throttling responses. Maybe not in the first iteration, but let's file a bug/GH issue to address it.

Reply:
will create issue